File: python/setMatrixZeroes.py
from re import T


def setZeroes(matrix: list[list[int]]) -> None:
    m = 0
    zeroes = []
    while m < len(matrix):
        n = 0
        while n < len(matrix[0]):
            if matrix[m][n] == 0:
                zeroes.append([m, n])
            n += 1
        m += 1

    for zero in zeroes:
        rowCol(matrix, zero[0], zero[1])


def rowCol(matrix: list[list[int]], m: int, n: int):
    start, end = 0, len(matrix)
    while start < end:
        matrix[start][n] = 0
        start += 1

    start, end = 0, len(matrix[0])
    while start < end:
        matrix[m][start] = 0
        start += 1


# space -> 0(1)
def setZeroesI(matrix: list[list[int]]) -> None:
    zeroRow = False
    zeroCol = False

    row, col = len(matrix), len(matrix[0])
    m, n = 0, 0
    while m < row:
        if matrix[m][0] == 0:
            zeroCol = True
            break
        m += 1

    while n < col:
        if matrix[0][n] == 0:
            zeroRow = True
            break
        n += 1

    m = 1
    while m < row:
        n = 1
        while n < col:
            if matrix[m][n] == 0:
                matrix[m][0] = 0
                matrix[0][n] = 0
            n += 1
        m += 1

    m = 1
    while m < row:
        if matrix[m][0] == 0:
            n = 1
            while n < col:
                matrix[m][n] = 0
                n += 1
        m += 1

    n = 1
    while n < col:
        if matrix[0][n] == 0:
            m = 1
            while m < row:
                matrix[m][n] = 0
                m += 1
        n += 1

    if zeroRow:
        n = 0
        while n < col:
            matrix[0][n] = 0
            n += 1

    if zeroCol:
        m = 0
        while m < row:
            matrix[m][0] = 0
            m += 1


matrix = [
    [1, 1, 1],
    [1, 0, 1],
    [1, 1, 1],
]

matrix1 = [
    [0, 1, 0],
    [1, 1, 1],
    [1, 1, 1],
]

setZeroes(matrix)
setZeroesI(matrix1)
print(matrix)
print(matrix1)

File: go/SetMatrixZeroes/SetMatrixZeroes.go
package main

import "fmt"

func setZeros(matrix [][]int) {
	zeros := make([][]int, 0)

	m := 0
	for m < len(matrix) {
		n := 0
		for n < len(matrix[0]) {
			if matrix[m][n] == 0 {
				zeros = append(zeros, []int{m, n})
			}
			n++
		}
		m++
	}

	for _, zero := range zeros {
		rowCol(matrix, zero[0], zero[1])
	}
}

func rowCol(matrix [][]int, m, n int) {
	start, stop := 0, len(matrix)
	for start < stop {
		matrix[start][n] = 0
		start++
	}
	start, stop = 0, len(matrix[0])
	for start < stop {
		matrix[m][start] = 0
		start++
	}
}

// Space -> 0(1)
func setZerosI(matrix [][]int) {
	zeroRow := false
	zeroCol := false

	for m := 0; m < len(matrix[0]); m++ {
		if matrix[0][m] == 0 {
			zeroRow = true
			break
		}
	}

	for n := 0; n < len(matrix); n++ {
		if matrix[n][0] == 0 {
			zeroCol = true
			break
		}
	}

	for m := 1; m < len(matrix); m++ {
		for n := 1; n < len(matrix[0]); n++ {
			if matrix[m][n] == 0 {
				matrix[m][0] = 0
				matrix[0][n] = 0
			}
		}
	}

	for m := 1; m < len(matrix); m++ {
		if matrix[m][0] == 0 {
			for n := 1; n < len(matrix[0]); n++ {
				matrix[m][n] = 0
			}
		}
	}

	for n := 1; n < len(matrix[0]); n++ {
		if matrix[0][n] == 0 {
			for m := 1; m < len(matrix); m++ {
				matrix[m][n] = 0
			}
		}
	}

	if zeroRow {
		for m := 0; m < len(matrix[0]); m++ {
			matrix[0][m] = 0
		}
	}

	if zeroCol {
		for n := 0; n < len(matrix); n++ {
			matrix[n][0] = 0
		}
	}
}

func main() {
	matrix := [][]int{
		{1, 1, 1},
		{1, 0, 1},
		{1, 1, 1},
	}

	matrixI := [][]int{
		{0, 1, 0},
		{1, 1, 1},
		{1, 1, 1},
	}

	setZeros(matrix)
	setZerosI(matrixI)
	fmt.Println(matrix)
	fmt.Println(matrixI)
}
